mathics/builtin/layout.py

```diff
@@ -0,0 +1,846 @@
+from mathics.core.formatter import lookup_method as lookup_format_method
+from mathics.builtin.lists import list_boxes
+from mathics.builtin.makeboxes import MakeBoxes
+from mathics.builtin.options import options_to_rules
+from mathics.builtin.tensors import get_dimensions
+
+from mathics.core.atoms import (
+    Integer,
+    Real,
+    String,
+    StringFromPython,
+)
+
+from mathics.core.element import EvalMixin
+from mathics.core.expression import Expression, BoxError
+from mathics.core.list import ListExpression
+from mathics.core.symbols import (
+    Symbol,
+    SymbolFullForm,
+    SymbolList,
+)
+
+from mathics.core.systemsymbols import (
+    SymbolMakeBoxes,
+    SymbolRowBox,
+    SymbolRule,
+)
+
+MULTI_NEWLINE_RE = re.compile(r"\n{2,}")
+
+SymbolNumberForm = Symbol("System`NumberForm")
+SymbolSuperscriptBox = Symbol("System`SuperscriptBox")
+SymbolSubscriptBox = Symbol("System`SubscriptBox")
+
+
+class Row(Builtin):
+    """
+    <dl>
+    <dt>'Row[{$expr$, ...}]'
+        <dd>formats several expressions inside a 'RowBox'.
+    </dl>
+    """
+
+    summary_text = "1D layouts containing arbitrary objects in a row"
+
+    def apply_makeboxes(self, items, sep, f, evaluation):
+        """MakeBoxes[Row[{items___}, sep_:""],
+        f:StandardForm|TraditionalForm|OutputForm]"""
+
+        items = items.get_sequence()
+        if not isinstance(sep, String):
+            sep = MakeBoxes(sep, f)
+        if len(items) == 1:
+            return MakeBoxes(items[0], f)
+        else:
+            result = []
+            for index, item in enumerate(items):
+                if index > 0 and not sep.sameQ(String("")):
+                    result.append(to_boxes(sep, evaluation))
+                item = MakeBoxes(item, f).evaluate(evaluation)
+                item = to_boxes(item, evaluation)
+                result.append(item)
+            return RowBox(*result)
+
+
+class Grid(Builtin):
+    """
+    <dl>
+    <dt>'Grid[{{$a1$, $a2$, ...}, {$b1$, $b2$, ...}, ...}]'
+        <dd>formats several expressions inside a 'GridBox'.
+    </dl>
+
+    >> Grid[{{a, b}, {c, d}}]
+     = a   b
+     .
+     . c   d
+    """
+
+    options = GridBox.options
+    summary_text = " 2D layout containing arbitrary objects"
+
+    def apply_makeboxes(self, array, f, evaluation, options) -> Expression:
+        """MakeBoxes[Grid[array_?MatrixQ, OptionsPattern[Grid]],
+        f:StandardForm|TraditionalForm|OutputForm]"""
+        return GridBox(
+            ListExpression(
+                *(
+                    ListExpression(
+                        *(
+                            Expression(SymbolMakeBoxes, item, f)
+                            for item in row.elements
+                        ),
+                    )
+                    for row in array.elements
+                ),
+            ),
+            *options_to_rules(options),
+        )
+
+
+SymbolTableDepth = Symbol("TableDepth")
+
+
+class TableForm(Builtin):
```

Comment:
WIth respect to XXXForm. https://reference.wolfram.com/search/?q=Form lists a little under 6K forms 

If we are aspire to do something coming anywhere close to that. We can't have an organization like this where we put in one module a bunch of individual forms. Probably more like it would be a module to hold various kinds of forms would be a start. And a straightforward way in which we and other can add more Forms.

Reply:
At this moment, what I am trying to do is to disentangle the interface of the `BaseElement` subclasses from the `MakeBoxes` loop, without producing another thing highly entangled with the core. Once we have that, we could add more formats as pymathics pluggins.

Reply:
That makes sense. Just keep it in mind, though.

Reply:
And also note that we will have to provide the basic set of Forms. And even for that, this we should not try to list all of the built-in Forms in one module (mixed with other stuff). 

Should also mention with respect to pymathics that it's namespacing scoping should be fixed up.  And down the line we will probably want a real "autoload" like GNU Emacs has.